File: examples/algebraic-data-types/algebraic-data-types.kt
sealed class Shape {
  data class Square(
    val side: Double
  ) : Shape()
  data class Circle(
    val radius: Double
  ) : Shape()
}

fun getArea(shape: Shape): Double {
  return when (shape) {
    is Shape.Square ->
      shape.side * shape.side
    is Shape.Circle ->
      kotlin.math.PI * shape.radius * shape.radius
  }
}

fun main() {
  val shapes = listOf(
    Shape.Square(2.0),
    Shape.Circle(4.0)
  )
  val totalArea = shapes.fold(0.0) {
    sum, shape -> sum + getArea(shape)
  }
  println(totalArea)
}

File: examples/algebraic-data-types/algebraic-data-types.rs
enum Shape {
  Square { side: f64 },
  Circle { radius: f64 },
}

fn get_area(shape: &Shape) -> f64 {
  return match shape {
    Shape::Square { side } => side * side,
    Shape::Circle { radius } => {
      std::f64::consts::PI * radius * radius
    }
  };
}

fn main() {
  let shapes: Vec<Shape> = vec![
    Shape::Square { side: 2.0 },
    Shape::Circle { radius: 4.0 },
  ];

  let total_area =
    shapes.iter().fold(0f64, |sum, shape| {
      sum + get_area(&shape)
    });

  print!("{:?}", total_area)
}
